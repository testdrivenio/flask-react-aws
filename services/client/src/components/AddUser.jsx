import React from "react";
import PropTypes from "prop-types";
import { Formik } from "formik";
import * as Yup from "yup";

import "./form.css";

const AddUser = props => (
  <Formik
    initialValues={{
      username: "",
      email: "",
      password: ""
    }}
    onSubmit={(values, { setSubmitting, resetForm }) => {
      props.addUser(values);
      resetForm();
      setSubmitting(false);
    }}
    validationSchema={Yup.object().shape({
      username: Yup.string()
        .required("Username is required.")
        .min(6, "Username must be greater than 5 characters."),
      email: Yup.string()
        .email("Enter a valid email.")
        .required("Email is required.")
        .min(6, "Email must be greater than 5 characters."),
      password: Yup.string()
        .required("Password is required.")
        .min(11, "Password must be greater than 10 characters.")
    })}
  >
    {props => {
      const {
        values,
        touched,
        errors,
        isSubmitting,
        handleChange,
        handleBlur,
        handleSubmit
      } = props;
      return (
        <form onSubmit={handleSubmit}>
          <div className="field">
            <label className="label" htmlFor="input-username">
              Username
            </label>
            <input
              name="username"
              id="input-username"
              className={
                errors.username && touched.username ? "input error" : "input"
              }
              type="text"
              placeholder="Enter a username"
              value={values.username}
              onChange={handleChange}
              onBlur={handleBlur}
            />
            {errors.username && touched.username && (
              <div className="input-feedback">{errors.username}</div>
            )}
          </div>
          <div className="field">
            <label className="label" htmlFor="input-email">
              Email
            </label>
            <input
              name="email"
              id="input-email"
              className={
                errors.email && touched.email ? "input error" : "input"
              }
              type="email"
              placeholder="Enter an email address"
              value={values.email}
              onChange={handleChange}
              onBlur={handleBlur}
            />
            {errors.email && touched.email && (
              <div className="input-feedback">{errors.email}</div>
            )}
          </div>
          <div className="field">
            <label className="label" htmlFor="input-password">
              Password
            </label>
            <input
              name="password"
              id="input-password"
              className={
                errors.password && touched.password ? "input error" : "input"
              }
              type="password"
              placeholder="Enter a password"
              value={values.password}
              onChange={handleChange}
              onBlur={handleBlur}
            />
            {errors.password && touched.password && (
              <div className="input-feedback">{errors.password}</div>
            )}
          </div>
          <input
            type="submit"
            className="button is-primary"
            value="Submit"
            disabled={isSubmitting}
          />
        </form>
      );
    }}
  </Formik>
);

AddUser.propTypes = {
  addUser: PropTypes.func.isRequired
};

export default AddUser;
