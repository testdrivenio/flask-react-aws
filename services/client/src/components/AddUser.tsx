import React from "react";
import axios from "axios";
import { Formik, Form, Field } from "formik";
import { z } from "zod";
import {
  Box,
  Button,
  FormControl,
  FormLabel,
  Input,
  Heading,
} from "@chakra-ui/react";
import { FormErrorMessage } from "@chakra-ui/react";

interface UserObject {
  username: string;
  email: string;
  password?: string;
  created_date?: string;
}

interface AddUserProps {
  addUserToList: (user: UserObject) => void;
}

type FieldProps = {
  field: {
    name: string;
    value: string;
    onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
    onBlur: (event: React.FocusEvent<HTMLInputElement>) => void;
  };
};

const validationSchema = z.object({
  username: z
    .string()
    .min(4, "Username must be at least 4 characters long")
    .regex(
      /^[a-zA-Z0-9_]+$/,
      "Username can only contain letters, numbers, and underscores",
    ),
  email: z.string().email("Enter a valid email"),
  password: z.string().min(6, "Password must be at least 6 characters long"),
});

type FormValues = z.infer<typeof validationSchema>;

const AddUser: React.FC<AddUserProps> = ({ addUserToList }) => {
  const handleSubmit = async (
    values: FormValues,
    {
      setSubmitting,
      resetForm,
    }: {
      setSubmitting: (isSubmitting: boolean) => void;
      resetForm: () => void;
    },
  ) => {
    try {
      const response = await axios.post(
        `${import.meta.env.VITE_API_SERVICE_URL}/users`,
        values,
      );

      if (response.status === 201) {
        console.log(response.data.message);

        const newUser = {
          username: values.username,
          email: values.email,
          created_date: new Date().toISOString(),
        };

        addUserToList(newUser);
        resetForm();
      }
    } catch (error) {
      console.error("There was an error registering the user:", error);
    } finally {
      setSubmitting(false);
    }
  };

  return (
    <Box p={4} maxW="1200px" mx="auto" mt={12}>
      <Box mb={6} textAlign="left">
        <Heading as="h2" size="xl" color="gray.700">
          Register a User
        </Heading>
      </Box>

      <Formik
        initialValues={{ username: "", email: "", password: "" }}
        validate={(values) => {
          try {
            validationSchema.parse(values);
            return {};
          } catch (error) {
            return (error as z.ZodError).formErrors.fieldErrors;
          }
        }}
        onSubmit={handleSubmit}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form>
            <Box
              p={4}
              borderWidth={1}
              borderRadius="lg"
              boxShadow="lg"
              maxW="500px"
              textAlign="left"
            >
              <Field name="username">
                {({ field }: FieldProps) => (
                  <FormControl
                    isRequired
                    mb={6}
                    isInvalid={!!(errors.username && touched.username)}
                  >
                    <FormLabel fontSize="lg" htmlFor="input-username">
                      Username
                    </FormLabel>
                    <Input
                      {...field}
                      id="input-username"
                      placeholder="Enter a username"
                    />
                    {errors.username && touched.username && (
                      <FormErrorMessage>{errors.username}</FormErrorMessage>
                    )}
                  </FormControl>
                )}
              </Field>

              <Field name="email">
                {({ field }: FieldProps) => (
                  <FormControl
                    isRequired
                    mb={6}
                    isInvalid={!!(errors.email && touched.email)}
                  >
                    <FormLabel fontSize="lg" htmlFor="input-email">
                      Email
                    </FormLabel>
                    <Input
                      {...field}
                      id="input-email"
                      type="email"
                      placeholder="Enter an email address"
                    />
                    {errors.email && touched.email && (
                      <FormErrorMessage>{errors.email}</FormErrorMessage>
                    )}
                  </FormControl>
                )}
              </Field>

              <Field name="password">
                {({ field }: FieldProps) => (
                  <FormControl
                    isRequired
                    mb={6}
                    isInvalid={!!(errors.password && touched.password)}
                  >
                    <FormLabel fontSize="lg" htmlFor="input-password">
                      Password
                    </FormLabel>
                    <Input
                      {...field}
                      id="input-password"
                      type="password"
                      placeholder="Enter a password"
                    />
                    {errors.password && touched.password && (
                      <FormErrorMessage>{errors.password}</FormErrorMessage>
                    )}
                  </FormControl>
                )}
              </Field>

              <Button
                type="submit"
                colorScheme="green"
                bg="green.400"
                size="lg"
                width="full"
                isLoading={isSubmitting}
              >
                Submit
              </Button>
            </Box>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default AddUser;
