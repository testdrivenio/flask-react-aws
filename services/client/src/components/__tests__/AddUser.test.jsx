import React from 'react';
import { render, cleanup } from '@testing-library/react';

import AddUser from '../AddUser';

afterEach(cleanup);

const props = {
  addUser: () => { return true },
}

it('renders with default props', () => {
  const { getByLabelText, getByText } = render(<AddUser {...props} />);

  const usernameInput = getByLabelText('Username');
  expect(usernameInput).toHaveAttribute('type', 'text');
  expect(usernameInput).not.toHaveValue();

  const emailInput = getByLabelText('Email');
  expect(emailInput).toHaveAttribute('type', 'email');
  expect(emailInput).not.toHaveValue();

  const passwordInput = getByLabelText('Password');
  expect(passwordInput).toHaveAttribute('type', 'password');
  expect(passwordInput).not.toHaveValue();

  const buttonInput = getByText('Submit');
  expect(buttonInput).toHaveValue('Submit');
});

it("renders", () => {
  const { asFragment } = render(<AddUser {...props} />);
  expect(asFragment()).toMatchSnapshot();
});
