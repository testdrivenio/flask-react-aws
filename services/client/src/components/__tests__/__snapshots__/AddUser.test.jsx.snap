// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <form>
    <div
      class="field"
    >
      <label
        class="label"
        for="input-username"
      >
        Username
      </label>
      <input
        class="input"
        id="input-username"
        name="username"
        placeholder="Enter a username"
        type="text"
        value=""
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="input-email"
      >
        Email
      </label>
      <input
        class="input"
        id="input-email"
        name="email"
        placeholder="Enter an email address"
        type="email"
        value=""
      />
    </div>
    <div
      class="field"
    >
      <label
        class="label"
        for="input-password"
      >
        Password
      </label>
      <input
        class="input"
        id="input-password"
        name="password"
        placeholder="Enter a password"
        type="password"
        value=""
      />
    </div>
    <input
      class="button is-primary"
      type="submit"
      value="Submit"
    />
  </form>
</DocumentFragment>
`;
